### PR TITLE
Simd enablement

### PR DESCRIPTION
**Pull request by shawnl**
Not in scope:
* new memory model for comptime
* stack allocate vectors, so that addr-of vectors (&a[3]) can work, like they are supported in gcc-9

Some stuff is blocked on #447.

While this mostly follows #903, there are some decisions that had to be made, the biggest one is support both bit-wise and bool operations on bools, because a() and b() is differn't from a() & b(), as if a() returns true the first will not run b() and the second will. This is explained in the commit.

---
TODO
* needs more tests of zero-length vectors
* block exporting incompatible ABIs: 256-bit wide on pre-avx, and 512-bit wide on pre-avx512

### CODE REVIEW

**Comment by dimenus** [comment]
@shawnl thoughts on horizontal add?

**Comment by shawnl** [comment]
Some discussion on horizontal intrinsics in LLVM here https://github.com/ziglang/zig/issues/903#issuecomment-515591534 which effect #2698

File: std/math/fma.zig [hunk 1]
```diff
@@ -13,6 +14,7 @@ pub fn fma(comptime T: type, x: T, y: T, z: T) T {
     return switch (T) {
         f32 => fma32(x, y, z),
         f64 => fma64(x, y, z),
+        f128 => fma128(x, y, z),
```
Comment (data-man) [inline std/math/fma.zig]: Where is ```fma128``` implementation?

**Comment by data-man** [comment]
```zig
var v: u32 = 5;
var x = @splat(4, v);
```
Can be changed to
```zig
var v: u32 = 5;
var x = @Vector(4, v);
```
?

File: doc/langref.html.in [hunk 2]
```diff
@@ -7650,7 +7650,7 @@ fn List(comptime T: type) type {
       {#header_open|@truncate#}
       <pre>{#syntax#}@truncate(comptime T: type, integer: var) T{#endsyntax#}</pre>
       <p>
-      This function truncates bits from an integer type, resulting in a smaller
+      This function truncates bits from an integer type (or vector of integers), resulting in a smaller
```
Comment (daurnimator) [inline doc/langref.html.in]: It's not obvious if you mean it truncates the vector as a whole, or each element in a vector.

File: src/all_types.hpp [hunk 3]
```diff
@@ -2527,20 +2534,31 @@ struct IrInstructionBinOp {
...
 
     IrInstruction *ptr;
 };
 
+struct IrInstructionExtractInsert {
+    IrInstruction base;
+
+    IrInstruction *agg;
+    IrInstruction *index;
+    // nullptr if Extract
```
Comment (daurnimator) [inline src/all_types.hpp]: Looks like this doesn't need to be separate fields?

File: src/codegen.cpp [hunk 4]
```diff
@@ -4102,6 +4191,49 @@ static LLVMValueRef ir_render_ctz(CodeGen *g, IrExecutable *executable, IrInstru
     return gen_widen_or_shorten(g, false, int_type, instruction->base.value.type, wrong_size_int);
 }
 
+static LLVMValueRef ir_render_shuffle_vector(CodeGen *g, IrExecutable *executable, IrInstructionShuffleVector *instruction) {
+    uint64_t len_a = instruction->a->value.type->data.vector.len,
+        len_c = instruction->mask->value.type->data.vector.len;
```
Comment (daurnimator) [inline src/codegen.cpp]: For clarity I would make this an independent statement.

File: src/ir.cpp [hunk 5]
```diff
@@ -12734,6 +12925,30 @@ static IrInstruction *ir_analyze_cast(IrAnalyze *ira, IrInstruction *source_inst
         return ir_analyze_widen_or_shorten(ira, source_instr, value, wanted_type);
     }
 
+    // widening of vectors
+    // These are separate (while identical) as I am still not sure if this should not implicitely cast,
+    // but only explicitely cast. (i.e. with @cast, or @as, or @Vector(4, i32)(foo)),
```
Comment (daurnimator) [inline src/ir.cpp]: Does this need deciding on before merge?

Also, typo: "explicitly"

File: src/ir_print.cpp [hunk 6]
```diff
@@ -396,7 +396,18 @@ static void ir_print_return_ptr(IrPrint *irp, IrInstructionReturnPtr *instructio
     fprintf(irp->f, "@ReturnPtr");
 }
 
-static void ir_print_load_ptr(IrPrint *irp, IrInstructionLoadPtr *instruction) {
+static void ir_print_extractinsert(IrPrint *irp, IrInstructionExtractInsert *instruction) {
```
Comment (daurnimator) [inline src/ir_print.cpp]: Should this have an underscore between `extract` and `insert`?

File: test/stage1/behavior/shuffle.zig [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+            // Upcasting of a
+            var v3: @Vector(2, i32) = [2]i32{ 2147483647, -2};
+            const mask4: @Vector(4, i32) = [4]i32{ 0, ~i32(2), 1, ~i32(3)};
+            res = @shuffle(i32, v3, x, mask4);
+            expect(mem.eql(i32, ([4]i32)(res), [4]i32{ 2147483647, 3, -2, 4 }));
+        }
+    };
+    S.doTheTest();
+    comptime S.doTheTest();
+}
```
Comment (daurnimator) [inline test/stage1/behavior/shuffle.zig]: missing newline at EOF.

File: doc/langref.html.in (hunk 2 above)
Comment (shawnl) [reply doc/langref.html.in]: good point

File: src/all_types.hpp (hunk 3 above)
Comment (shawnl) [reply src/all_types.hpp]: No, these extract fields, and all the others, are used for Extract

Comment (shawnl) [reply src/all_types.hpp]: No, these extract fields, and all the others, are used for Extract.

File: src/codegen.cpp (hunk 4 above)
Comment (shawnl) [reply src/codegen.cpp]: Yeah

**Comment by shawnl** [comment]
@data-man I don't like that because the capital letter `V` means *type*.

File: std/math/fma.zig (hunk 1 above)
Comment (shawnl) [reply std/math/fma.zig]: whoops. I was working on it, and that accidentally god added.

**Comment by shawnl** [comment]
ewwwwwwwwwwwww, github sorts commits by date, rather than ordering.

File: doc/langref.html.in [hunk 8]
```diff
@@ -7997,11 +7997,21 @@ pub const TypeInfo = union(TypeId) {
...
       {#syntax#}undefined{#endsyntax#} {#syntax#}ElemType{#endsyntax#} vector with length of {#syntax#}mask{#endsyntax#}.</p>
       <p>
-      {#syntax#}ElemType{#endsyntax#} must be an {#link|integer|Integers#}, a {#link|float|Floats#}, or a
+      {#syntax#}ElemType{#endsyntax#} must be an {#link|integer|Integers#}, a {#link|bool|Primitive Types#}, a {#link|float|Floats#}, or a
       {#link|pointer|Pointers#}. The mask may be any vector length that the target supports, and its' length determines the result length.
       </p>
       {#header_close#}
 
+      {#header_open|@splat#}
+      <pre>{#syntax#}@shuffle(comptime len: u32, scalar: var) @Vector(len, @typeOf(scalar)){#endsyntax#}</pre>
```
Comment (dbandstra) [inline doc/langref.html.in]: ^ typo? (shuffle instead of splat)

File: test/stage1/behavior/vector.zig [hunk 9]
```diff
@@ -80,3 +110,539 @@ test "array to vector" {
...
+            expect(y[1] == 0);
+            expect(y[2] == 1);
+            expect(y[3] == 0);
+        }
+    };
+    S.doTheTest();
+    comptime S.doTheTest();
+}
+
+test "vectors - indxed load from vector, SIMD" {
```
Comment (data-man) [inline test/stage1/behavior/vector.zig]:
```suggestion
test "vectors - indexed load from vector, SIMD" {
```

File: test/stage1/behavior/vector.zig [hunk 10]
```diff
@@ -632,3 +632,17 @@ test "arrays - indexed SIMD store" {
...
+        fn doTheTest() void {
+            var v = [_]u32{7, 8};
+            var s: @Vector(2, usize) = [_]usize{1, 0};
+            var res = v[s];
+            expect(res[0] == 8);
+            expect(res[1] == 7);
+        }
+    };
+    S.doTheTest();
+    // comptime S.doTheTest();
```
Comment (shawnl) [inline test/stage1/behavior/vector.zig]: Perhaps I should just remove this, as bounds checks only make sense for run-time code.

**Comment by andrewrk** [comment]
I have merged the following commits into master (the first 3 commits in this pull request):

 * fb8e95c3f5d902e0f0564a2ee0eb55c24e86d257
 * 9ce1fe275121cb5e69d3d94477e2a2503211b66f
 * acdf18226244ab47572095250dfa0517170bcf97

You are welcome to chop those off and rebase against master, or continue to do nothing while I go through the commits one by one. I haven't hit conflicts yet. Probably best to double check with me on IRC before doing any work to make sure we aren't clobbering each other.

**Comment by andrewrk** [comment]
I have merged into master:

 * 9e0e5d3dd54fc5f8c027c4fac3d11c9d79bc76fd

**Comment by andrewrk** [comment]
I merged 20d89dfe29dba39d9e514d6a828d1889fd14494d into master. I skipped over a couple more difficult commits.

**Comment by andrewrk** [comment]
@shawnl can you rebase now? This patch set is at least a bit smaller now, with some of the commits merged into master. Each commit is taking quite a bit of fixups, so you'll have to work with me as we chip away at this.

**Comment by andrewrk** [comment]
By the way I accepted the `@as` proposal (#1757).

**Comment by shawnl** [comment]
Here is the example that works in gcc-9:
```
typedef unsigned char v16uc __attribute__ ((vector_size (16)));

#include <stdio.h>
#include <stdint.h>
void pr(unsigned char *a) {
printf("%hhu %zx\n", *a, (uintptr_t)a);
}

int main() {
  v16uc v = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15};
  pr(&v[3]);
}
```
and a corresponding zig test case:
```
pub fn main() void {
  var v: @Vector(16, u8) = [_]u8{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15};
  pointerToU8(&v[3]);
}
fn pointerToU8(x: *u8) void {
  @import("std").debug.warn("{} {x}\n", x.*, x);
}
```

**Comment by andrewrk** [comment]
@shawnl - 

I really want to get this merged. I looked into this today and it's a bit overwhelming to try to do all the commits at once. Here's my plan for how this can work, if you're willing to meet me halfway:

 * Keep this PR open
 * At the same time, use it to create 1 new PR that only does 1 self-contained, tested thing.
 * We work on that smaller PR until it is merged.
 * Repeat.

The first item would be **Array Accesses on Vectors**. If you open a PR which only does this, I think we can get it merged swiftly. It seems everything else in this PR depends on that feature.

Does that work for you?

**Comment by andrewrk** [comment]
> The first item would be **Array Accesses on Vectors**. If you open a PR which only does this, I think we can get it merged swiftly. It seems everything else in this PR depends on that feature.

See #3575 

This PR implements array access of vectors to mean element access, but it's planned for that to be how to dereference a vector of pointers.

I definitely am going to need this PR to be split into distinct smaller mergeable pieces.

**Comment by shawnl** [comment]
Yes I'm getting on it.

However I disagree with this proposal. I don't see what is wrong with the
way gcc-9 does it. I find your way confusing and unnecessarily different
from use of arrays.

As you discovered we can also make it work for odd-bit-widths if we change
llvm's bit packing of vectors (when written out) to pad like arrays are (I
don't really think this is a good idea however).

**Comment by shawnl** [comment]
The lowering to `@shuffle`, `@gather`, and `@scatter` of loads and stores got dropped while re-basing this.

**Comment by andrewrk** [comment]
This PR won't be merged, but I'm definitely interested in the smaller PRs that bring these commits in 1 at a time. I'll leave it up to @shawnl to track this fork and do the work of slowly upstreaming it.
